app/auth/__init__.py

```diff
@@ -1,68 +1,132 @@
 import traceback
 from functools import wraps
-from flask import jsonify, request
+from flask import jsonify, request, abort
 from flask_jwt_simple import JWTManager, create_jwt, get_jwt_identity, jwt_required
 from jwt.exceptions import PyJWTError
 from flask_jwt_simple.exceptions import InvalidHeaderError, NoAuthorizationError
-from auth.ldap import LdapAuthenticator
+
 from app import logger
-from utils.env_loader import config
+from auth.ldap import LdapAuthenticator
+from auth.exceptions import ApplicationUserRoleException
+from auth.authenticator import EmptyAuthenticator
+from models import db
+from models.user import User
+from models.application import Application
+from models.application_user_role import ApplicationUserRole, Role
 
 
 class Auth(object):
-    enabled = False
-
-    def __init__(self, app=None):
-        if app is not None:
-            self.init_app(app)
+    __enabled = False
 
     @staticmethod
     def auth_required(fn):
+        def check_role(user_id, application_id, method):
+            role = db.session.query(ApplicationUserRole).filter(
+                ApplicationUserRole.application_id == application_id,
+                ApplicationUserRole.user_id == user_id).one_or_none()
+            if role is None:
+                # applications which don't have users are also accesssible as owner
+                roles = db.session.query(ApplicationUserRole).filter(
+                    ApplicationUserRole.application_id == application_id).count()
+                if roles == 0:
+                    return True
+                return False
+            if method == 'GET':
+                return True
+            elif role.role == Role.editor or role.role == Role.owner:
+                    return True
+            return False
+
         @wraps(fn)
         def wrapper(*args, **kwargs):
-            if Auth.enabled and request.path.startswith('/api/') and not request.path.startswith('/api/settings') and not request.path.startswith('/api/kubernetes/dump'):
-                return jwt_required(fn(*args, **kwargs))
+            if Auth.__enabled and request.path.startswith('/api/') and not request.path.startswith('/api/settings') and not request.path.startswith('/api/kubernetes/dump'):
+                @jwt_required
+                def run():
+                    application_id = kwargs.get('application_id')
+                    if application_id is not None:
+                        user_id = get_jwt_identity()
+                        if not check_role(user_id, application_id, request.method):
+                            raise ApplicationUserRoleException
+                    return fn(*args, **kwargs)
+                return run()
             else:
                 return fn(*args, **kwargs)
         return wrapper
 
-    def init_app(self, app, api, **kwargs):
-        Auth.enabled = True
+    @staticmethod
+    def enabled(enabled=None):
```

Comment:
I think this method should be separated.
`is_enabled()` and `set_enabled()`.

Reply:
fixed.